dd-sdk-android/src/main/kotlin/com/datadog/android/log/internal/domain/LogGenerator.kt

```diff
@@ -0,0 +1,101 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.log.internal.domain
+
+import com.datadog.android.core.internal.net.info.NetworkInfoProvider
+import com.datadog.android.log.LogAttributes
+import com.datadog.android.log.internal.user.UserInfoProvider
+import com.datadog.android.rum.GlobalRum
+import io.opentracing.util.GlobalTracer
+
+internal class LogGenerator(
+    internal val serviceName: String,
+    internal val loggerName: String,
+    internal val networkInfoProvider: NetworkInfoProvider?,
+    internal val userInfoProvider: UserInfoProvider,
+    envName: String,
+    appVersion: String
+) {
+
+    private val envTag: String? = if (envName.isNotEmpty()) {
+        "${LogAttributes.ENV}:$envName"
+    } else {
+        null
+    }
+
+    private val appVersionTag = if (appVersion.isNotEmpty()) {
+        "${LogAttributes.APPLICATION_VERSION}:$appVersion"
+    } else {
+        null
+    }
+
+    @Suppress("LongParameterList")
+    fun generateLog(
+        level: Int,
+        message: String,
+        throwable: Throwable?,
+        attributes: Map<String, Any?>,
+        tags: Set<String>,
+        timestamp: Long,
+        threadName: String? = null,
+        bundleWithTraces: Boolean = true,
+        bundleWithRum: Boolean = true
+    ): Log {
+        val combinedAttributes = resolveAttributes(attributes, bundleWithTraces)
+        val combinedTags = resolveTags(tags, bundleWithRum, combinedAttributes)
+        return Log(
+            serviceName = serviceName,
+            level = level,
+            message = message,
+            timestamp = timestamp,
+            throwable = throwable,
+            attributes = combinedAttributes,
+            tags = combinedTags.toList(),
+            networkInfo = networkInfoProvider?.getLatestNetworkInfo(),
+            userInfo = userInfoProvider.getUserInfo(),
+            loggerName = loggerName,
+            threadName = threadName ?: Thread.currentThread().name
+        )
+    }
+
+    private fun resolveTags(
+        tags: Set<String>,
+        bundleWithRum: Boolean,
+        combinedAttributes: MutableMap<String, Any?>
+    ): MutableSet<String> {
+        val combinedTags = mutableSetOf<String>().apply { addAll(tags) }
+        envTag?.let {
+            combinedTags.add(it)
+        }
+        appVersionTag?.let {
+            combinedTags.add(it)
+        }
+        if (bundleWithRum && GlobalRum.isRegistered()) {
+            val activeContext = GlobalRum.getRumContext()
+            combinedAttributes[LogAttributes.RUM_APPLICATION_ID] = activeContext.applicationId
+            combinedAttributes[LogAttributes.RUM_SESSION_ID] = activeContext.sessionId
+            combinedAttributes[LogAttributes.RUM_VIEW_ID] = activeContext.viewId
+        }
```

Comment:
This block should be in the `resolveAttributes()` method.

Reply:
Damn...you're totally right.